posts/2024-09-28-duckdb_polars_reshape-toby_hocking/index.qmd

```diff
@@ -0,0 +1,584 @@
+
+## Creating `part` and `dim` columns
+
+Both `polars` and `duckdb` are not capable of producing the separate
+`part` and `dim` columns during the reshape operation, but we can
+always do it as a post-processing step. One way to do that, by
+specifying a separator, would be via `data.table::tstrsplit`, as in
+the code below:
+
+```{r}
+data.table(iris.long.duckdb)[
+, c("part","dim") := tstrsplit(part_dim, split="[.]")
+][]
+```
+
+Another way of doing that, by specifying a regex, would be via
+`nc::capture_first_df` (recently given the `data.table` [Seal of
+Approval](https://rdatatable-community.github.io/The-Raft/posts/2024-08-01-seal_of_approval-nc/)),
+as in the code below:
+
+```{r}
+nc::capture_first_df(iris.long.duckdb, part_dim=list(
+  part=".*",
+  "[.]",
+  dim=".*"))
+```
+
+Both results above are data tables with extra cols `part` and `dim`. For visualization, these data tables could be used with either `facet_grid` or `facet_wrap`, similar to the examples above.
+
+## Reshape into multiple columns
+
+Another kind of wide-to-long reshape involves reshaping into multiple columns. 
+For example, in the iris data, we may wonder whether sepals are larger than petals (in terms of both length and width). 
+To answer that question, we could make a scatterplot of `y=Sepal` versus `x=Petal`, with a facet/panel for each dimension (`Length` and `Width`). 
+In the ggplot system, we would need to compute a data table with columns `Sepal`, `Petal`, and `dim`, and we can do that by specifying the `value.name` keyword to `measure()`, as in the code below:
+
+```{r}
+(iris.long.parts <- melt(iris.wide, measure.vars=measure(value.name, dim, sep=".")))
+```
+
+Again, the `measure()` function in the code above operates by
+splitting the input column names using `sep`, which results in two
+groups (`Sepal` and `Width`, etc) for each of the measured
+columns. The `value.name` keyword indicates that each unique value in
+the first group (`Sepal` and `Petal`) should be used as the name of an
+output column.
+This functionality can be very convenient for some data reshaping tasks, but it is neither supported `polars`, nor in `duckdb`. 
+Going back to our original motivating problem, we can make the scatterplot using the code below,
+
+```{r iris-scatter}
+ggplot()+
+  theme_bw()+
+  geom_abline(slope=1, intercept=0, color="grey")+
+  geom_point(aes(
+    Petal, Sepal),
+    data=iris.long.parts)+
+  facet_grid(. ~ dim, labeller=label_both)+
+  coord_equal()
+```
+
+From the plot above, we see that since all of the data points (black) are above the y=x line (grey), so we can conclude that sepals are indeed larger than petals, in terms of both length and width.
```

Comment:
Given the use of since, I'd omit 'so':

From the plot above, we see that since all of the data points (black) are above the y=x line (grey), we can conclude ...

Reply:
or I'd rather omit 'since':

'From the plot above, we see that all of the data points (black) are above the y=x line (grey), so we can conclude ... '